### ISSUE TITLE
Include Cleaner: support IWYU pragma: export

### ISSUE THREAD

**Issue by stk-ableton** [clangd/clangd#1085]
We have a number of "umbrella headers" in our code base; these are headers that don't provide anything themselves, but only include other headers which are not supposed to be included directly. Including an umbrella header results in an "unused include" warning.

As far as I can see, the only way to silence these warnings right now is to add `// IWYU pragma: keep` at the include site, which works but is not ideal because it has to be done in hundreds of places.

It would be nicer to be able to add `// IWYU pragma: export` (or `begin_exports`/`end_exports`) to the includes inside the umbrella header, so that this only has to be done in one place.

**Comment by sam-mccall** [comment]
We support the inverse: `// IWYU pragma: private; include "foo.h"`. Is it possible to use that instead?

The `export` pragma is a bit unfortunately designed: when the parser sees a symbol's definition the pragma may or may not be visible. (in particular, when you open the private header directly!). And it's also possible for multiple headers to claim ownership of the same header.

Nevertheless, maybe we can add support, or at least say something like "never mark a header as unused if it has any export pragmas".

**Comment by kirillbobyrev** [comment]
Oops, I was late with my reply.

The summary of what I wanted to say is this: we looked through some of the most popular pragmas. The most useful ones seem to be `keep`, `private`, `export`/`begin_exports` and `no_include`. As you mentioned, we have `keep` support now and the `IWYU pragma private: include "public.h"` is in the works ([D120306](https://reviews.llvm.org/D120306)). I expect to land the `private` pragma patch quite soon, so this would hopefully solve the issue you're describing.

As Sam mentioned, there's ambiguity with `export` pragmas and we decided to leave it out of the initial implementation (so far we don't have the production-ready version). I think it's doable but just has the lower priority and we already support `private` but not `export` for incude insertion (in code competion).

**Comment by sam-mccall** [comment]
Ah sorry, indeed I meant to say we supporte `private` today for code completion, and soon for include-cleaner.

**Comment by stk-ableton** [comment]
`private` is not going to work for us, unfortunately, for two reasons:

- Sometimes the private headers are third-party headers that we would prefer not to modify
- Other times the umbrella header is just provided as a convenience, but clients are also allowed to include the more specific ones if they prefer

> "never mark a header as unused if it has any export pragmas".

Not really great, but probably better than nothing, so if this is easy to do, it would be greatly appreciated.

**Comment by kirillbobyrev** [comment]
> `private` is not going to work for us, unfortunately, for two reasons:
> 
> * Sometimes the private headers are third-party headers that we would prefer not to modify

Yes, this is also something we're looking into. Third-party is a problem, because realistically when you don't have any control over it it might be too optimistic to expect the pragmas to be correct. We want to make it possible to filter the warnings based on the header paths to disable Include Cleaner for `third_party/` internally. That would also be critical for automated clean-ups that we would like to do (see https://discourse.llvm.org/t/rfc-lifting-include-cleaner-missing-unused-include-detection-out-of-clangd/61228).

> * Other times the umbrella header is just provided as a convenience, but clients are also allowed to include the more specific ones if they prefer

Correct me if I'm wrong, but I was under the impression that this is not how `export` pragma works. AFAIK from the IWYU standpoint (and Include Cleaner for that matter) that actually means that the exporter is __responsible__ for the symbols of the private header. This model is quite simple and doesn't introduce unexpected results.

For the case you described to work as intended we would also have to make changes to the _policy_ and make it configurable. We're likely to build some configuration mechanism and it might be one of the options to support but this really seems like a niche use-case.

> 
> > "never mark a header as unused if it has any export pragmas".
> 
> Not really great, but probably better than nothing, so if this is easy to do, it would be greatly appreciated.

This being said, we would be happy to support more pragmas and have some flexibility as we understand that different users would want different behavior. However, right now the top priority is driving the tooling to the production state so that we can have at least one consistent policy that would be correct in most cases. Once we reach that, we'll surely start expanding the feature set. Also, right now the code structure is changing rapidly so it might be a bit harder to plug more pragmas in (partially just because it's not totally obvious where some of them would fit in), and reusing the code that we have for include insertion helps us get the prototype running quickly.

This is all to say that contributions are certainly welcome: we might not have the capacity to support the exact use-case you described right now but we'd be happy to have it once the code restructuring is complete.

**Comment by stk-ableton** [comment]
> We want to make it possible to filter the warnings based on the header paths to disable Include Cleaner for `third_party/` internally.

Will this help for our concrete situation though? The way you describe it, it sounds like this would only work if the umbrella header is also part of the third-party code. In our case, the umbrella header is our code, outside of `third_party/`.

> Correct me if I'm wrong, but I was under the impression that this is not exactly how `export` pragma is intended to work. AFAIK from the IWYU standpoint (and Include Cleaner for that matter) `export` actually means that the umbrella header is **responsible** for providing the symbols from the private header.

I have no idea to be honest, I haven't used IWYU myself and don't know how it works. From reading the docs, it was my impression that `export` is treated as if the included header is inlined at that point. In my mental model this should make it possible to include the exported header in several different umbrella headers, and for the client to include any of those without getting a warning. But I'm probably missing a lot about how Include Cleaner works internally.

**Comment by sam-mccall** [comment]
I think you're right. export isn't really the inverse of private.
If you have
  - foo.h: `int foo();`
  - umbrella1.h: `#include "foo.h" // export`
  - umbrella2.h: `#include "foo.h" // export`

Then we could say any of {foo.h, umbrella1.h, umbrella2.h} are valid to include. (If it were the inverse of private, then foo.h would not be valid).
In include-cleaner, supporting this would achieve:
 - suppress false `unused include` warnings for umbrella headers
 - suppress false `missing include` warnings for foo.h when an umbrella is included
 - allow includes of umbrellas to be suggested as a fix, if they're in the translation unit

It would not tell us how to rank/filter suggested includes, whereas private can help with that.
(This is the *only* question include-insertion cares about, which explains why we historically ignored this pragma).

I think we should probably work out how to support this in include-cleaner.

**Comment by sliedes** [comment]
There seems to be at least some support for this nowadays. The documentation currently says:

> IWYU tool offers a set of [pragmas](https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/IWYUPragmas.md). Include Cleaner respects keep, always_keep, private, and export/begin_exports/end_exports.

Is something missing from this support? If so, probably the documentation should be updated; if not, probably this issue can be closed?

**Comment by kirillbobyrev** [comment]
> There seems to be at least some support for this nowadays. The documentation currently says:
> 
> > IWYU tool offers a set of [pragmas](https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/IWYUPragmas.md). Include Cleaner respects keep, always_keep, private, and export/begin_exports/end_exports.
> 
> Is something missing from this support? If so, probably the documentation should be updated; if not, probably this issue can be closed?

Good point, I think those were added a long while ago now, but the issue wasn't updated.

[Documentation](https://clangd.llvm.org/design/include-cleaner#iwyu-pragmas) clearly mentions it now, so it should be 100% covered by now.

I'm closing the issue, feel free to reopen if something you wanted still isn't working/covered by the current implementation.

**Comment by stk-ableton** [comment]
(Coming back to this after a bit of delay, sorry for that...)
(It doesn't look like I can reopen the issue, so I'm commenting on it anyway hoping that people will see this.)

With clangd 18.1.3 (or with current snapshots) our umbrella header issue is _not_ solved. It was solved with clangd 16.x, but I suppose this was because it implemented the simpler "never mark a header as unused if it has any export pragmas" logic.

Our scenario is this:

```
main.cpp:
    #include "umbrella.hpp"

    int main(void) {
      someFunction();
      return 0;
    }

umbrella.hpp:
    #pragma once
    #include "third-party/umbrella.hpp" // IWYU pragma: export

third-party/umbrella.hpp:
    #pragma once
    #include "detail.hpp"

third-party/detail.hpp:
    #pragma once
    void someFunction();
```

With 18.1.3 this warns in main.cpp that umbrella.hpp is unused; with 16.x it doesn't.

Note that our own umbrella.hpp is the only header that we can add IWYU pragmas to.

While playing with this I noticed that the warning also goes away when I remove the `#pragma once` from `third-party/detail.hpp`. Obviously this is not a solution, but maybe it's still interesting.
